### DISCUSSION TITLE
Implementing iDEAL and Giropay payment methods with Checkout.com

### DISCUSSION THREAD

**Discussion by sudhirrnanda** [juspay/hyperswitch#4142] [General]
### Context 

Embracing diversity in payments is one of the core tenets of Hyperswitch. And we look forward to integrating a large variety of payment processors (connectors) and payment flows around the world.

Checkout.com is a global payment company that provides a gateway, an acquirer and processor as part of its end-to-end solution. On Hyperswitch, Checkout.com is one of the connectors we have already integrated with. As of now, Hyperswitch supports payment methods such as Cards, Applepay and GooglePay through Checkout.com. 

We look forward to including some more critical payment methods through Checkout.com

### Problem Statement 
iDeal and Giropay are online banking based payment methods that allow end users to complete payments in their own banking environment. 

- iDEAL is a leading payment method in the Netherlands with 62%+ market share and processes close to 1 Billion transactions annually. 
- Giropay is a popular payment method in Germany, and supports about 32,000 online businesses. 

**The goal is to extend Hyperswitch’s integration with Checkout.com to support iDeal and Giropay.**

### How to get started

1. [Set up a Rust environment and run Hyperswitch locally](https://github.com/juspay/hyperswitch/blob/main/docs/try_local_system.md) 

2. [Signup and creating a sandbox account on Checkout.com](http://checkout.com/) to get the API Credentials 

### How to Implement:  

Now, you can get started with implementation: 

1. Read and understand the [general guidelines](https://github.com/juspay/hyperswitch/blob/main/docs/CONTRIBUTING.md) on contributing to Hyperswitch

2. Import postman collection of hyperswitch from [here](https://www.postman.com/hyperswitch/workspace/hyperswitch-development/collection/25176162-630b5353-7002-44d1-8ba1-ead6c230f2e3) 

3. Understand the implementation of all different payment flows and [how to add a connector](https://github.com/juspay/hyperswitch/blob/main/add_connector.md) 

4. As mentioned earlier for Checkout.com, card payments are implemented. Test out this flow from postman using these [test resources](https://www.checkout.com/docs/developer-resources/testing/test-cards) and understand the API implementation 

5. Study the relevant [API documentation of Checkout.com](http://checkout.com/) to implement iDeal and Giropay through Checkout.com

6. Implement iDeal and Giropay using the API references in point 5. Modify the following file to implement iDeal and Giropay
          

> crates/router/src/connector/checkout/transformers.rs
>           crates/router/src/connector/checkout.rs
> 

 
7. Test the following flows - Payment, Capture, Void and Refund using postman for your  local implementation and generate relevant screenshots of the test cases. 

### Expected Outcome: 

Submit a **pull request** to the [juspay/hyperswitch](https://github.com/juspay/hyperswitch) repository including the relevant screenshots of the test cases. 

_Brownie Points:_ Providing clear and comprehensive documentation and code comments.

Note: We welcome multiple PRs for this problem statement

### More Useful Resources: 
[Setting up Checkout.com on Hyperswitch](http://checkout.com/)

**Comment by Rupesh-0001** [comment]
I have raised PR https://github.com/juspay/hyperswitch/pull/4177/ for this and have implemented GiroPay, I am working on the iDEAL part. Can you please assign this issue to me?

**Reply by manojradhakrishnan** [reply]
Hi @Rupesh-0001 You may create an issue on [juspay/hyperswitch](https://github.com/juspay/hyperswitch) and assign it to yourself. For the time-being, the PR reference would suffice.

**Reply by Rupesh-0001** [reply]
Hi @manojradhakrishnan thank you for your reply, I've created issue https://github.com/juspay/hyperswitch/issues/4181 but I am unable to assign myself or link my PR to it.

**Comment by 26tanishabanik** [comment]
I tried setting up the dev environment following these steps: https://github.com/juspay/hyperswitch/blob/main/docs/try_local_system.md#set-up-a-development-environment-using-docker-compose 
But I am getting this error upon running `curl --head --request GET 'http://localhost:8080/health'`:
`curl: (56) Recv failure: Connection reset by peer`

@manojradhakrishnan , can you help me with this ?

**Reply by prajjwalkumar17** [reply]
Hey @26tanishabanik can you share the server logs, so that we can help you out.

**Comment by Garv-M** [comment]
Is Giropay or ideal implemented on Hyperswitch on a diffrent router? @manojradhakrishnan

**Reply by swangi-kumari** [reply]
Yes @Garv-M , 
Giropay or ideal are implemented for other processors such as Adyen, Stripe and Mollie.

**Reply by Garv-M** [reply]
Fine, Thank you :)

**Comment by varsubham** [comment]
hello team, I tried dev setup using docker-compose following this doc https://github.com/juspay/hyperswitch/blob/main/docs/try_local_system.md#set-up-a-development-environment-using-docker-compose

 but getting this error, can you please help if this looks familiar ?

> hyperswitch-migration_runner-1 exited with code 0
hyperswitch-pg-1                  | 2024-04-03 09:18:02.108 UTC [61] LOG:  checkpoint starting: time
hyperswitch-hyperswitch-server-1  | error: could not compile `router` (lib)
hyperswitch-hyperswitch-server-1  |
hyperswitch-hyperswitch-server-1  | Caused by:
hyperswitch-hyperswitch-server-1  |   process didn't exit successfully: `/usr/local/rustup/toolchains/1.77.1-aarch64-unknown-linux-gnu/bin/rustc --crate-name router --edition=2021 crates/router/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no -C debuginfo=2 --cfg 'feature="accounts_cache"' --cfg 'feature="backwards_compatibility"' --cfg 'feature="business_profile_routing"' --cfg 'feature="connector_choice_mca_id"' --cfg 'feature="default"' --cfg 'feature="dummy_connector"' --cfg 'feature="frm"' --cfg 'feature="kv_store"' --cfg 'feature="olap"' --cfg 'feature="oltp"' --cfg 'feature="payout_retry"' --cfg 'feature="payouts"' --cfg 'feature="profile_specific_fallback_routing"' --cfg 'feature="retry"' --cfg 'feature="stripe"' -C metadata=de90fc565da7705e -C extra-filename=-de90fc565da7705e --out-dir /cargo_build_cache/debug/deps -C incremental=/cargo_build_cache/debug/incremental -L dependency=/cargo_build_cache/debug/deps --extern actix_cors=/cargo_build_cache/debug/deps/libactix_cors-4267aac48ef6aa7e.rmeta --extern actix_http=/cargo_build_cache/debug/deps/libactix_http-151de0e84060ce3e.rmeta --extern actix_multipart=/cargo_build_cache/debug/deps/libactix_multipart-e3aa6253198454fe.rmeta --extern actix_rt=/cargo_build_cache/debug/deps/libactix_rt-a3dff3643ca58d86.rmeta --extern actix_web=/cargo_build_cache/debug/deps/libactix_web-2c0a73317a097907.rmeta --extern analytics=/cargo_build_cache/debug/deps/libanalytics-a68e12828eac21fd.rmeta --extern api_models=/cargo_build_cache/debug/deps/libapi_models-452203c4dea70cf1.rmeta --extern argon2=/cargo_build_cache/debug/deps/libargon2-9574ec7cf417d2af.rmeta --extern async_bb8_diesel=/cargo_build_cache/debug/deps/libasync_bb8_diesel-2ea188e58799efab.rmeta --extern async_trait=/cargo_build_cache/debug/deps/libasync_trait-7719a0a5f2bf15a6.so --extern base64=/cargo_build_cache/debug/deps/libbase64-a83631b74c058dc3.rmeta --extern bb8=/cargo_build_cache/debug/deps/libbb8-8710e5e998cec157.rmeta --extern bigdecimal=/cargo_build_cache/debug/deps/libbigdecimal-34d5c1c8d768659c.rmeta --extern blake3=/cargo_build_cache/debug/deps/libblake3-3aef09b4000a37a6.rmeta --extern bytes=/cargo_build_cache/debug/deps/libbytes-06acf229c5c58aa5.rmeta --extern cards=/cargo_build_cache/debug/deps/libcards-ab6de439e44d530c.rmeta --extern clap=/cargo_build_cache/debug/deps/libclap-391d78e3aaa4b34b.rmeta --extern common_enums=/cargo_build_cache/debug/deps/libcommon_enums-f20ec5c7ad30831d.rmeta --extern common_utils=/cargo_build_cache/debug/deps/libcommon_utils-8077027f4f1f44f3.rmeta --extern config=/cargo_build_cache/debug/deps/libconfig-5ae123dd1091ebc6.rmeta --extern cookie=/cargo_build_cache/debug/deps/libcookie-3cde1642af827263.rmeta --extern currency_conversion=/cargo_build_cache/debug/deps/libcurrency_conversion-8e7c9fc8c2adef6d.rmeta --extern data_models=/cargo_build_cache/debug/deps/libdata_models-d48fe7340df16637.rmeta --extern diesel=/cargo_build_cache/debug/deps/libdiesel-c9c8c6d0827781dd.rmeta --extern diesel_models=/cargo_build_cache/debug/deps/libdiesel_models-007535ff85b71793.rmeta --extern digest=/cargo_build_cache/debug/deps/libdigest-ca1ad3535d0abe95.rmeta --extern dyn_clone=/cargo_build_cache/debug/deps/libdyn_clone-dbc897de00b1b01e.rmeta --extern encoding_rs=/cargo_build_cache/debug/deps/libencoding_rs-d166423304a6575a.rmeta --extern erased_serde=/cargo_build_cache/debug/deps/liberased_serde-832f184786903bdd.rmeta --extern error_stack=/cargo_build_cache/debug/deps/liberror_stack-8d30fa84a456274e.rmeta --extern euclid=/cargo_build_cache/debug/deps/libeuclid-c89cf343badd49ff.rmeta --extern external_services=/cargo_build_cache/debug/deps/libexternal_services-fe919f6ffe2ff247.rmeta --extern futures=/cargo_build_cache/debug/deps/libfutures-83d797dce42bead6.rmeta --extern hex=/cargo_build_cache/debug/deps/libhex-5f0db979c983eb24.rmeta --extern http=/cargo_build_cache/debug/deps/libhttp-f99b50b63df4a751.rmeta --extern hyper=/cargo_build_cache/debug/deps/libhyper-26da7c424134509d.rmeta --extern hyperswitch_interfaces=/cargo_build_cache/debug/deps/libhyperswitch_interfaces-bcede08068caeea8.rmeta --extern image=/cargo_build_cache/debug/deps/libimage-2a2ac71191100225.rmeta --extern infer=/cargo_build_cache/debug/deps/libinfer-97819cd588dea140.rmeta --extern iso_currency=/cargo_build_cache/debug/deps/libiso_currency-6ba7ef138dfa9566.rmeta --extern isocountry=/cargo_build_cache/debug/deps/libisocountry-a0d94bdfc00d4e9d.rmeta --extern josekit=/cargo_build_cache/debug/deps/libjosekit-cdd3c137f83eb7dc.rmeta --extern jsonwebtoken=/cargo_build_cache/debug/deps/libjsonwebtoken-c5560f423e0adf3c.rmeta --extern kgraph_utils=/cargo_build_cache/debug/deps/libkgraph_utils-4bb8b9633f96f9ea.rmeta --extern masking=/cargo_build_cache/debug/deps/libmasking-67765c7692958fb7.rmeta --extern maud=/cargo_build_cache/debug/deps/libmaud-9d302c1790959e1c.rmeta --extern mime=/cargo_build_cache/debug/deps/libmime-faff8333f02573ac.rmeta --extern nanoid=/cargo_build_cache/debug/deps/libnanoid-87d6ff64397b84f7.rmeta --extern num_cpus=/cargo_build_cache/debug/deps/libnum_cpus-4a6d5901c9904b01.rmeta --extern once_cell=/cargo_build_cache/debug/deps/libonce_cell-649eda0d8d54cafe.rmeta --extern openssl=/cargo_build_cache/debug/deps/libopenssl-5d2c03ff90f5e2c3.rmeta --extern pm_auth=/cargo_build_cache/debug/deps/libpm_auth-f51e34bce89a7998.rmeta --extern qrcode=/cargo_build_cache/debug/deps/libqrcode-68169341479aa408.rmeta --extern quick_xml=/cargo_build_cache/debug/deps/libquick_xml-2033df9e7d2307f4.rmeta --extern rand=/cargo_build_cache/debug/deps/librand-202cb1f9ecb3cb60.rmeta --extern rand_chacha=/cargo_build_cache/debug/deps/librand_chacha-7313053a60ec3496.rmeta --extern rdkafka=/cargo_build_cache/debug/deps/librdkafka-3cd069d849a147fb.rmeta --extern redis_interface=/cargo_build_cache/debug/deps/libredis_interface-ccaee7ee6bddcf7d.rmeta --extern regex=/cargo_build_cache/debug/deps/libregex-5b44ca8cafa5bd42.rmeta --extern reqwest=/cargo_build_cache/debug/deps/libreqwest-a4b5d16d74236347.rmeta --extern ring=/cargo_build_cache/debug/deps/libring-d975df411e3ae0ff.rmeta --extern router_derive=/cargo_build_cache/debug/deps/librouter_derive-30eb941342bbec3f.so --extern router_env=/cargo_build_cache/debug/deps/librouter_env-91678d1bae1b63c9.rmeta --extern roxmltree=/cargo_build_cache/debug/deps/libroxmltree-b31910f12c712e24.rmeta --extern rust_decimal=/cargo_build_cache/debug/deps/librust_decimal-754eb66ce9aa4d48.rmeta --extern rustc_hash=/cargo_build_cache/debug/deps/librustc_hash-57f5a41b28fa2f6e.rmeta --extern scheduler=/cargo_build_cache/debug/deps/libscheduler-4caf53da3db0acc5.rmeta --extern serde=/cargo_build_cache/debug/deps/libserde-ba7f01fca2cda49c.rmeta --extern serde_json=/cargo_build_cache/debug/deps/libserde_json-8c2b4971030e556e.rmeta --extern serde_path_to_error=/cargo_build_cache/debug/deps/libserde_path_to_error-2040eb799fe1bc12.rmeta --extern serde_qs=/cargo_build_cache/debug/deps/libserde_qs-7d43b3ba36c9a724.rmeta --extern serde_urlencoded=/cargo_build_cache/debug/deps/libserde_urlencoded-3f9ab8d7c621038e.rmeta --extern serde_with=/cargo_build_cache/debug/deps/libserde_with-b297a5810f6996f7.rmeta --extern sha1=/cargo_build_cache/debug/deps/libsha1-0c2bf662ac663b84.rmeta --extern sqlx=/cargo_build_cache/debug/deps/libsqlx-590d11d9a4e3073f.rmeta --extern storage_impl=/cargo_build_cache/debug/deps/libstorage_impl-e5013d524093c00f.rmeta --extern strum=/cargo_build_cache/debug/deps/libstrum-e803fb1ecf31fe3b.rmeta --extern tera=/cargo_build_cache/debug/deps/libtera-05a959637a6274dc.rmeta --extern thiserror=/cargo_build_cache/debug/deps/libthiserror-3cd571fc68bde6a1.rmeta --extern time=/cargo_build_cache/debug/deps/libtime-88e9e92aaf166072.rmeta --extern tokio=/cargo_build_cache/debug/deps/libtokio-b0b7a2dbcee33f65.rmeta --extern tracing_futures=/cargo_build_cache/debug/deps/libtracing_futures-6dbb89aad140b4ce.rmeta --extern unicode_segmentation=/cargo_build_cache/debug/deps/libunicode_segmentation-1588d55b8c007512.rmeta --extern url=/cargo_build_cache/debug/deps/liburl-475574045025a103.rmeta --extern utoipa=/cargo_build_cache/debug/deps/libutoipa-05ea658f2fe426da.rmeta --extern uuid=/cargo_build_cache/debug/deps/libuuid-b4df1b212745d68d.rmeta --extern validator=/cargo_build_cache/debug/deps/libvalidator-f75749488d805150.rmeta --extern x509_parser=/cargo_build_cache/debug/deps/libx509_parser-2d958068a1bfe543.rmeta -Funsafe_code '-Aclippy::option_map_unit_fn' '-Wclippy::as_conversions' '-Wclippy::expect_used' '-Wclippy::index_refutable_slice' '-Wclippy::indexing_slicing' '-Wclippy::match_on_vec_items' '-Wclippy::missing_panics_doc' '-Wclippy::out_of_bounds_indexing' '-Wclippy::panic' '-Wclippy::panic_in_result_fn' '-Wclippy::panicking_unwrap' '-Wclippy::todo' '-Wclippy::unimplemented' '-Wclippy::unreachable' '-Wclippy::unwrap_in_result' '-Wclippy::unwrap_used' '-Wclippy::use_self' -Wrust_2018_idioms -Wunused_qualifications -L native=/cargo_build_cache/debug/build/zstd-sys-a0946edc8c50e0d3/out -L native=/usr/lib/aarch64-linux-gnu -L native=/cargo_build_cache/debug/build/ring-c2596b69def220de/out -L native=/cargo_build_cache/debug/build/ring-53697b67f1e30750/out -L native=/cargo_build_cache/debug/build/blake3-f289566f691ca102/out -L native=/cargo_build_cache/debug/build/rdkafka-sys-1b1d7c05c546b4d7/out/src` (signal: 9, SIGKILL: kill)
hyperswitch-hyperswitch-server-1 exited with code 101
hyperswitch-pg-1                  | 2024-04-03 09:18:38.803 UTC [61] LOG:  checkpoint complete: wrote 352 buffers (2.1%); 1 WAL file(s) added, 0 removed, 0 recycled; write=36.532 s, sync=0.057 s, total=36.699 s; sync files=583, longest=0.002 s, average=0.001 s; distance=3472 kB, estimate=3472 kB; lsn=0/1C771C0, redo lsn=0/1C77188

**Reply by TakshPanchal** [reply]
I faced the same error, checkout this [discussion](https://github.com/juspay/hyperswitch/discussions/4145#discussioncomment-8932172). 
My laptop has 8 GB RAM so, I followed [this](https://github.com/juspay/hyperswitch/blob/main/docs/try_local_system.md#set-up-a-development-environment-using-docker-compose) did all the setup locally, **without** using docker.

**Reply by varsubham** [reply]
Thanks @TakshPanchal working fine without using docker

**Comment by Subh513** [comment]
Hi team , I am currently in the process of integrating Giropay payments through checkout.com .I have encountered a couple of challenges
 Also checkout.com requires organization email for Sandbox account setup.


Issue 1: processing_channel_id_invalid: When attempting to initiate a payment request through the sandbox API for Giropay, I received an error indicating that the processing channel ID is invalid. I understand that the processing channel ID is essential for routing payments correctly. Could you kindly provide guidance on how to obtain the correct processing_channel_id for Giropay payments?

Issue 2: Payee Not Onboarded: Additionally, I encountered an error stating that the payee is not onboarded. I believe this may be related to the configuration of my Checkout.com account. however i created my test account in checkout.com . i have Public key ,Secret key .  How to onboard in  to use and test this api ?

<img width="1007" alt="Screenshot 2024-04-07 at 1 31 10 AM" src="https://github.com/juspay/hyperswitch/assets/166251977/6d121a73-ccb8-418d-a913-42c22a5ec316">
<img width="1007" alt="Screenshot 2024-04-07 at 1 30 20 AM" src="https://github.com/juspay/hyperswitch/assets/166251977/8e90ee97-2bf7-47ee-b1d0-8e48aa2a08e5">

**Reply by SrikantKonduri** [reply]
1. You can find `processing_channel_id` in your checkout dashboard under keys section
2. I am also stuck here and according to checkout documentation we have to mail them for integration
check this https://www.checkout.com/docs/payments/add-payment-methods/ideal

**Reply by Subh513** [reply]
Thanks , i am able  to get but after adding processing_channel_id i am getting response as payee_not_onboarded attaching the screenshot below .
i guess we have to mail them i.e support team to checkout 
<img width="1007" alt="Screenshot 2024-04-07 at 1 13 52 PM" src="https://github.com/juspay/hyperswitch/assets/166251977/61606f1f-e892-45b5-8681-5e81aa2e951b">

**Reply by SrikantKonduri** [reply]
Yes , mail them and wait for them to reply

**Comment by sid-rh** [comment]
I cannot create a checkout test account without providing a business email.
